### ISSUE TITLE
Potential issue in function khash_put_iovec()

### ISSUE THREAD

**Issue by mehoffer** [systemd/systemd#17646]
The blow initialization of fields ``msg_iov`` and ``msg_iovlen`` doesn't appear to be what was actually intended I guess

https://github.com/systemd/systemd/blob/e6ce785d611702532ae71547e78fbb49eff3aded/src/basic/khash.c#L249-L253

Wasn't the intention here to use _Designated Initializers_ and have something like this?

```c
         struct msghdr mh = { 
                 .msg_iov = (struct iovec*) iovec, 
                 .msg_iovlen = n, 
         }; 
```

As a result, the original implementation actually tends to initialize two first fields of the ``msghdr`` structure seen below i.e. ``msg_name`` and ``msg_namelen`` respectively (at least, this is what would happen when compiling it with clang). 

```c
struct msghdr
  {
    void *msg_name;             /* Address to send to/receive from.  */
    socklen_t msg_namelen;      /* Length of address data.  */

    struct iovec *msg_iov;      /* Vector of data to send/receive into.  */
    size_t msg_iovlen;          /* Number of elements in the vector.  */
```

This can also be observed by enabling ``missing-field-initializers`` and inspecting the warning being produced:
```
../src/basic/khash.c:253:9: warning: missing initializer for field 'msg_iov' of 'struct msghdr' [-Wmissing-field-initializers]
```

One additional note about the function ``khash_put_iovec`` itself is that it doesn't seem to be called from anywhere in the codebase so just INFO... but still I think would be good to either fix it or just remove if above analysis turned out to be correct.

**Comment by yuwata** [comment]
Thanks. Fix is waiting in #17652.

> One additional note about the function khash_put_iovec itself is that it doesn't seem to be called from anywhere in the codebase so just INFO...

Actually, many functions in khash.c are not used. But, I'd like to keep them. From the commit message in 0fe5f3c5d743a7e4c63580a67066935f9e23a2f4, 
> In the long run we should try to remove the dependency on libgcrypt, in favour
    of using only the kernel's own primitives, however this is unlikely to happen
    anytime soon, as the kernel does not provide Elliptic Curve APIs to userspace
    at this time, and we need them for the DNSSEC cryptographic.

So, possibly, these function will be used in the future. I guess these unused function does not significantly increase the library size.

**Comment by mehoffer** [comment]
Sounds reasonable. Thanks.
